feat: Implement X509 Cert Based Authentication

vverman opened the pull request:
## 

With this PR, we are now able to use X509 Certificates to authenticate via Google Auth Library.

## Testing
Added unit tests and all integration tests passing as per [this](https://docs.google.com/document/d/1uutxffv21zZfZb1FVIaReROrWh_jiZS7dtB9NFZh_NI/edit?tab=t.0#heading=h.jjf3ex8pq9ke) testing document

## Additional Information

Change also includes trust chain verification to be passed to STS endpoint

vverman committed:
Included initial interfaces and options for creating x509client.

vverman committed:
Added implementation for x509provider

vverman committed:
Augmented logic for well-known cert config.

vverman committed:
Added changes to create CertificateSubjectTokenSupplier

vverman committed:
Added feature to call STS endpoint with the leaf certificate as trust chain.

vverman committed:
Added logic to use trust chains.

vverman committed:
Cleaned up certificateSubjectTokenSupplier and added mtlsClientTransporter logic to IdentityPoolClients Transporter

vverman committed:
Added tests for certificateConfig type externalClient

vverman committed:
All x509 auth logic in src/auth/certificatesubjecttokensupplier.ts

nbayati commented on src/auth/certificatesubjecttokensupplier.ts:
```diff
@@ -0,0 +1,299 @@
+   * @returns A promise that resolves with the subject token.
+   */
+  public async getSubjectToken(): Promise<string> {
+    const opts: GaxiosOptions = {method: 'GET'};
```
Are these two lines needed?

nbayati commented on src/auth/certificatesubjecttokensupplier.ts:
```diff
@@ -0,0 +1,299 @@
+   * @returns A promise that resolves with the subject token.
+   */
+  public async getSubjectToken(): Promise<string> {
+    const opts: GaxiosOptions = {method: 'GET'};
+    AuthClient.setMethodName(opts, 'getSubjectToken');
```
Are these two lines needed?

nbayati commented on src/auth/certificatesubjecttokensupplier.ts:
```diff
@@ -0,0 +1,299 @@
+import {X509Certificate} from 'crypto';
+import * as https from 'https';
+
+// --- Constants ---
```
I noticed the use of sectioning comments like `// --- Main Class ---`. While these can be useful for organization, it seems this pattern isn't common in the rest of our codebase. I think it's important to maintain consistency in the codebase.

nbayati commented on src/util.ts:
```diff
@@ -251,3 +258,43 @@ export function removeUndefinedValuesInObject(object: {[key: string]: any}): {
+/**
+ * Helper to check if a path points to a valid file.
+ */
+export function isValidFile(filePath: string): boolean {
```
Consider adding unit tests for the new helper methods.

vverman replied:
Done!

nbayati commented on src/auth/certificatesubjecttokensupplier.ts:
```diff
@@ -0,0 +1,299 @@
+   */
+  certificateConfigLocation?: string;
+  /**
+   * The file path to the trust chain (PEM format). If not provided,
+   * only the leaf certificate will be sent.
```
I think it's best if we remove the "If not provided, only the leaf certificate will be sent." part and just note that this field is optional. How we are creating the subject token to send to STS is an implementation detail that I think should not be exposed here, as it'd make the reader more confused than informed.

nbayati commented on src/auth/certificatesubjecttokensupplier.ts:
```diff
@@ -0,0 +1,299 @@
+   */
+  public createMtlsHttpsAgent(): https.Agent {
+    try {
+      const cert = fs.readFileSync(this.certPath);
+      const key = fs.readFileSync(this.keyPath);
+      return new https.Agent({key, cert});
+    } catch (err) {
+      const message = err instanceof Error ? err.message : String(err);
+      throw new CertificateSourceUnavailableError(
+        `Failed to read certificate or key file: ${message}`,
+      );
```
Suggestion: it might be helpful to include the file path in the error message. We could use two different try/catch blocks for reading the key and the certificate files, so we know which file caused the error.

nbayati commented on src/auth/identitypoolclient.ts:
```diff
@@ -87,6 +90,28 @@ export interface IdentityPoolClientOptions
+    certificate?: {
+      /**
+       * Specify whether the certificate config should be used from the default location
+       * either this or the certificate_config_location must be provided
```
nit: Please capitalize the first word and end each sentence with a period.

nbayati commented on src/auth/identitypoolclient.ts:
```diff
@@ -87,6 +90,28 @@ export interface IdentityPoolClientOptions
+      use_default_certificate_config?: boolean;
+      /**
+       * Location to fetch certificate config from in case default config is not to be used.
+       * either this or use_default_certificate_config=true should be provided
```
nit: Please capitalize the first word and end each sentence with a period.

nbayati commented on src/auth/identitypoolclient.ts:
```diff
@@ -87,6 +90,28 @@ export interface IdentityPoolClientOptions
+      /**
+       * Location to fetch trust chain from to send to STS endpoint.
+       * in case no location is provided, we will just send the leaf certificate as the
+       * trust chain
```
This is not really accurate. If a trust chain file is not provided, we just send the leaf certificate as the "subject token" and not the trust chain. I'd suggest focusing on what this field IS rather than the details of how we'll use it in the implementation. For instance, in the [go library](https://github.com/googleapis/google-cloud-go/blob/93c44c9e530005ba4e2bc83091625ed4b8c8604a/auth/credentials/externalaccount/externalaccount.go#L189-L197) we have the following comments for these three fields:

```
type CertificateConfig struct {
	// UseDefaultCertificateConfig, if true, attempts to load the default
	// certificate configuration. It checks the GOOGLE_API_CERTIFICATE_CONFIG
	// environment variable first, then a conventional default file location.
	// Cannot be true if CertificateConfigLocation is set.
	UseDefaultCertificateConfig bool `json:"use_default_certificate_config"`
	// CertificateConfigLocation specifies the path to the client certificate
	// and private key file. Must be set if UseDefaultCertificateConfig is
	// false or unset.
	CertificateConfigLocation string `json:"certificate_config_location"`
	// TrustChainPath specifies the path to a PEM-formatted file containing
	// the X.509 certificate trust chain. This file should contain any
	// intermediate certificates required to complete the trust chain between
	// the leaf certificate (used for mTLS) and the root certificate(s) in
	// your workload identity pool's trust store. The leaf certificate and
	// any certificates already present in the workload identity pool's trust
	// store are optional in this file. Certificates should be ordered with the
	// leaf certificate (or the certificate which signed the leaf) first.
	TrustChainPath string `json:"trust_chain_path"`
}
```

nbayati commented on src/auth/identitypoolclient.ts:
```diff
@@ -183,9 +209,30 @@ export class IdentityPoolClient extends BaseExternalAccountClient {
       } else {
         throw new Error(
-          'No valid Identity Pool "credential_source" provided, must be either file or url.',
+          'No valid Identity Pool "credential_source" provided, must be either file, or url, or certificate.',
```
nit: 
```suggestion
          'No valid Identity Pool "credential_source" provided, must be either file, url, or certificate.',
```

lsirac commented on test/test.identitypoolclient.ts:
Add tests:
* Malformed cert/key files: Tests for scenarios where the files pointed to by cert_path or key_path exist but are not valid PEM or key formats. 
* Malformed trust chain certs. While the order is tested, a test where one of the certificates within the trust chain PEM file is malformed should be added (this is related to the suggestion of more granular parsing I mentioned)
* Empty cert_config.json: Test the behavior when the cert_config.json file is empty or not valid JSON.
* Missing cert_path: Similar to the missing key_path test, add one for a missing cert_path in the config.

gcf-owl-bot[bot] committed:
chore(owlbot-nodejs): install 3.13.5 Python (#2042)

* chore: install higher version of Python

* chore: update to python 3.15

* update lagging dependency

* fix vulnerability

* change the version

Source-Link: https://github.com/googleapis/synthtool/commit/ca4c7ce65c001886c12b1c9b4ee216a7a1b807d2
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-nodejs:latest@sha256:6062c519ce78ee08490e7ac7330eca80f00f139ef1a241c5c2b306550b60c728

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

leahecole committed:
chore(deps): upgrade sinon to 21 (#2050)

* chore(deps): upgrade sinon to 21

* specify which timers to fake

* use @feywind's util for timers

* add crucial file

vverman commented on src/auth/certificatesubjecttokensupplier.ts:
That's a good question,

The synchronous file operations are needed because reading the certificate and key files are needed for creating the IdentityPoolClient. For e.g. to create the mtlsAgent. Since the constructor cannot be async, we need to do it synchronously.

I have changed the instances of file reads in processChainFromPaths() to async since that is called from the async getSubjectToken().

lsirac replied:
Can't we technically defer the creation of the mTLS `https.Agent` and the specific `Gaxios` transporter instance until they are about to be used with the token exchange methods, allowing the file reads to be done asynchronously?

vverman committed:
Added tests for malformed cert_config file, malfor
med certificate in trust chain.

renovate-bot committed:
fix(deps): update dependency @googleapis/iam to v30 (#2052)

gcf-owl-bot[bot] committed:
chore: add node 24 in node ci test (#2051)

Source-Link: https://github.com/googleapis/synthtool/commit/1218bc231201438192c962136303b95f0a11a4f5
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-nodejs:latest@sha256:66c44f0ad8f6caaa4eb3fbe74f8c2b4de5a97c2b930cee069e712c447723ba95

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

vverman committed:
Added unit tests for util

vverman commented on src/auth/identitypoolclient.ts:
```diff
@@ -87,6 +90,28 @@ export interface IdentityPoolClientOptions
        */
       subject_token_field_name?: string;
     };
+    /**
```
Done

lsirac commented on src/auth/baseexternalclient.ts:
```diff
@@ -715,4 +716,8 @@ export abstract class BaseExternalAccountClient extends AuthClient {
     return `gl-node/${nodeVersion} auth/${pkg.version} google-byoid-sdk source/${credentialSourceType} sa-impersonation/${saImpersonation} config-lifetime/${this.configLifetimeRequested}`;
   }
+
+  getTokenUrl(): string {
```
I believe in TypeScript members of a class (properties and methods) are public by default if no access modifier is explicitly specified. To make a method private, you need to use the private keyword. So let's mark this as protected.

lsirac commented on src/auth/baseexternalclient.ts:
```diff
@@ -234,12 +234,13 @@ export abstract class BaseExternalAccountClient extends AuthClient {
   public projectNumber: string | null;
   private readonly configLifetimeRequested: boolean;
   protected credentialSourceType?: string;
+  private readonly tokenUrl: string;
```
nit: maybe group these based on visibility

nbayati reviewed (approved):
LGTM, Good job! :)
